### PR TITLE
Fix endpoint selection for externalTrafficPolicy=local

### PR DESCRIPTION
**Pull request by ricky-rav**
In https://github.com/ovn-org/ovn-kubernetes/pull/4072 the following edge case was not handled correctly in ovnkube-controller code. 

Fix the case for `all endpoints terminating on a node when traffic policy is local`:
> "When the traffic policy is "Local" and all endpoints are terminating within a single node, then traffic should be routed to any terminating endpoint that is ready on that node."

https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/1669-proxy-terminating-endpoints/README.md#example-all-endpoints-terminating-on-a-node-when-traffic-policy-is-local

The endpoint selection logic in the services controller is now entirely implemented in `getEndpointsForService`, which computes for a given service and each service port all its cluster-wide endpoints and per-node local endpoints. We first apply a cluster-wide vs local endpoint selection and only then we apply readiness-based filtering with `getEligibleEndpointAddresses`.

Added unit tests for the new logic.

### CODE REVIEW

**Comment by ricky-rav** [comment]
I'll have to add a unit test for this particular case to `Test_buildPerNodeLBs`. I'll do that tomorrow morning.

**Comment by ricky-rav** [comment]
I've added some unit tests to `go-controller/pkg/ovn/controller/services/lb_config_test.go` that would have caught the issue I'm fixing here (ETP=local and terminating endpoints). @tssurya since you've done the work on ETP=local, would you be able to review those tests? Thanks!

File: go-controller/pkg/util/kube.go
Comment (jcaamano) [inline go-controller/pkg/util/kube.go:747]: I think it would work better as an overall strategy if you had a map of addresses to (selected) endpoint. Less iterations overall.

One iteration to build the map, as you do you zero out the endpoint address. Another iteration on the map, to populate the selected endpoint addresses.

Unless you require any kind of order stability, but that is not apparent otherwise I think it would be good to mention it.

File: go-controller/pkg/ovn/controller/services/utils.go
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/utils.go:12]: nit: fix naming & doc

File: go-controller/pkg/util/kube.go
Comment (jcaamano) [inline go-controller/pkg/util/kube.go:740]: This is also new right as we were not doing this kind of deduplication before? Does this have any relationship with the original issue you are trying to solve or is it a different issue altogether?

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/lb_config.go:63]: `zeroRouterLocalEndpoints` not used?

File: go-controller/pkg/util/kube_util.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube_util.go:1]: indeed I was wondering where to put it :)

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:740]: We were getting away with it by simply keeping a set (from the sets package) of IP addresses (https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/kube.go#L618-L619), since we knew we wanted them all, but now that I'm dealing with endpoints and not with IP addresses, I had to write the logic for it.

I realized I needed it because we have a unit test for it: https://github.com/ovn-org/ovn-kubernetes/blob/5f6cbb16f29239647b8db26d11c15dc33ca27455/go-controller/pkg/util/kube_test.go#L654

Then I looked at previous discussions in upstream kubernetes and saw that duplicate addresses are not removed in kube code and it's left to kube api clients to deal with them: https://github.com/kubernetes/kubernetes/issues/121556

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/lb_config.go:63]: ah ops, it's used in a downstream-only commit: https://github.com/openshift/ovn-kubernetes/blob/master/go-controller/pkg/ovn/controller/services/lb_config.go#L78
I'll remove it here.

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:747]: I'm totally with changing the logic, but I didn't get your approach :) 
Let's say I have an empty map of addresses to endpoints and I fill it in by iterating through all endpoints. The resulting values in the map will for sure include duplicate endpoints, since an endpoint can easily have more than one IP address. The main problem I faced is that the endpoints for an endpoint slice are not uniquely identified, it's just a list :-|

File: go-controller/pkg/util/kube.go
Comment (jcaamano) [reply go-controller/pkg/util/kube.go:750]: The fact that we had a set meant that we dedupped before as well, but what changed is the "how" regardless of the data structure:
```
we keep the IP address on the endpoint with the worse conditions (w.r.t ready, serving, terminating)
and remove the IP from the endpoint with the better conditions
```
Were we doing this with the set before?

File: go-controller/pkg/util/kube.go
Comment (jcaamano) [reply go-controller/pkg/util/kube.go]: It just can't happen because addressToEndpointIdx is only ever pupulated with indexes of endpoints which is not being changed.

Comment (jcaamano) [reply go-controller/pkg/util/kube.go]: In my good old days if we would get this paranoic we would add a panic

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:750]: We deduplicated also before, but without a criterion, since we were just handling IP addresses and not whole endpoints. Mind you, I'm also fine with keeping it simpler.

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go]: ok, let's not be paranoid then :D

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:740]: I like this solution!

One advantage is that I am setting `endpoints[i].Addresses` to an empty slice at the end of each iteration through the endpoints, so I don't need to remove any element from any slice. 

The implementation with the map of addresses to endpoint pointers would have easier to read, but in unit tests we expect a certain order of endpoints and, within endpoints, a certain order of addresses :| 

So I kept the endpoint ordering by dealing with indices instead of pointers as you suggested, but I'm a bit afraid that the unpredictable order of addresses within endpoints will yield flakes in our unit tests :|

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:733]: You're right, I didn't get that part either. I'll investigate on Monday.

Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:733]: I think that the deduplication is orthogonal to `PublishNotReadyAddresses`, since we want it to happen also when the service is configured to publish all addresses, I'll set it aside.

The point in the comment I linked is that a duplicate endpoint is not common and it's best to consider it as an error case that we resort to only if there's no better alternative. So If the same address appears in two endpoints, one of which is ready and the other of which is serving and terminating, we can be pessimistic about it and consider only the terminating and serving endpoint. We will use it only when we're falling back to terminating and serving endpoints and we will not mix it with ready endpoints. One can also argue that we may as well remove a duplicate endpoint all together if we consider it an error case, but the conversation can go on forever then :) 

As for this little paragraph:

> We already will choose terminating endpoints when no other endpoints exist. We chose at that time to NOT fall back on unready endpoints when no ready endpoints exist. If we want to revisit that, we can, but I don't think it's specific to this case.

I think the choice of words is confusing. I read "unready endpoints" as terminating & non-serving endpoints, otherwise the first sentence would be false and the whole KEP would break :)

Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:733]: And of course I'm fine with whatever decision. I had to make a decision in the code and I looked at what is done upstream to justify it.

File: go-controller/pkg/testing/kube.go
Comment (jcaamano) [inline go-controller/pkg/testing/kube.go:11]: Can't these be passed as parameters to the utility methods below instead of being defined as pseudo constants?

File: go-controller/pkg/ovn/controller/services/utils.go
Comment (jcaamano) [reply go-controller/pkg/ovn/controller/services/utils.go:12]: shouldn't it be more like `haveEndpointsAsHostAddress` but I see it backwards. Why not keep the original name?

File: go-controller/pkg/util/kube.go
Comment (jcaamano) [reply go-controller/pkg/util/kube.go:753]: I meant `else if otherEndpointIndex < len(endpoints) {`

File: go-controller/pkg/ovn/controller/services/utils.go
Comment (jcaamano) [reply go-controller/pkg/ovn/controller/services/utils.go:12]: I think I get it now. But it does proof the name is confusing :P

Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/utils.go:12]: How about `doEndpointsIncludeHostAddresses`?

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:734]: yes, I choose terminating and not serving over terminating and serving over ready.

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:747]: no, I wanted to make sure that addresses within an endpoint are unique before checking across endpoints.

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:776]: you're absolutely right. I'll add this bit to the unit tests, which is what I needed in the end.

Comment (ricky-rav) [reply go-controller/pkg/util/kube.go]: I moved the sorting of the IP addresses to the unit tests. As for the endpoint order, it comes for free when we go from `endpointsSet`to the list of endpoints...

File: go-controller/pkg/util/kube.go
Comment (jcaamano) [reply go-controller/pkg/util/kube.go:733]: I am just not sure I am making the same interpretation as you of the comment you are linking.

For example
> I don't think that's the common case, though. I think it's more likely that I have N ready and 1 (or small N) that is both ready and unready (dupplicates). In that case, optimism means we include an endpoint in a set which doesn't need it, and risk errors.

I understand here that duplicates should be excluded altogether vs your interpretation that duplicates should be picked through their worst endpoint.
So if i have IP A, B and duplicated C, I would just include A and B, and filter out C.
As in any information about C is un-trusted because it is duplicated.

That is as far as I can read into that comment.

It doesn't offer any hindsight on what should we do, for example, if C was the only IP and it was duplicated. And I don't have a clear picture in my head why the worst condition would be preferred over the best condition in this case.

File: go-controller/pkg/util/kube.go
Comment (jcaamano) [reply go-controller/pkg/util/kube.go]: Not really free if you have to do three ops vs one. So you have to iterate the map to create the set of indexes, then list them in order and then iterate that list; vs just iterating the map once to create the return list.

File: go-controller/pkg/util/kube.go
Comment (jcaamano) [reply go-controller/pkg/util/kube.go:733]: The overall logic might be fine. It is just that the implementation might make people read too much into it. It is more about "if we have a serving & not serving duplicated IP, we just exclude it because we don't know what to do with it; otherwise we include it, it doesn't really matter through which endpoint". If it does matter with what endpoint we include a duplicated IP, then that is the point I am missing.

Comment (jcaamano) [reply go-controller/pkg/util/kube.go:733]: There is also the question of how does this interpretation affect other functionality, like egress services and LB healthcheck.

File: go-controller/pkg/util/kube.go
Comment (ricky-rav) [reply go-controller/pkg/util/kube.go:746]: true, it would happen there anyway. I'll remove it.

**Comment by ricky-rav** [comment]
/hold 

Working on a new approach, as discussed with Jaime. Code is not clean yet and I'm working on the unit tests.

**Comment by ricky-rav** [comment]
@jcaamano as discussed, I implemented the new logic in one function, `getEndpointsForService`, that gets called once for a service and whose output is then parsed by the rest of the services controller to retrieve cluster-wide and per-node endpoint addresses. PTAL :)

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (jcaamano) [reply go-controller/pkg/ovn/controller/services/lb_config.go:46]: Also no need for `node`

Actually cluster endpoints and local endpoints are also part of lbconfig so it does look weird.

You could frame this as a free method
`func getNodeSwitchTargetV4IPs(node string, c *lbConfig) (targetIPs []string, isLocal bool)`

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/lb_config.go:875]: Does it make sense for this port attribute to be here, or rather in the `lbConfig` struct?

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/lb_config.go:874]: probably define this along with lbConfig for better visibility

**Comment by tssurya** [comment]
un-assigning myself here since Jaime is already doing a great job.

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/lb_config.go:122]: Sure! And the comment above is old. My bad!

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/lb_config.go:46]: Good idea! I made `getNodeSwitchTargetIPs` handle v4 and v6 at the same time. Does it look ok?
Also, do you prefer a free method because we're not really modifying `c *lbConfig`?

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/lb_config.go:875]: it's true that it's now duplicated in `clusterEndpoints` and `nodeEndpoints`. Let me see if I can move it up to `lbConfig`

**Comment by ricky-rav** [comment]
I've pushed a new `getEndpointsForService` as suggested in slack by Jaime. To reduce memory consumption, the per-node endpoints map is limited to the nodes in the local zone.

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/lb_config.go]: I was a bit afraid of taking this route. Can it wait?

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/lb_config.go:935]: We could initialize the capacities of these as well

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/lb_config.go:917]: I guess there is no issue with this but technically in the previous code we would only check this when assessing local endpoints, not for cluster endpoints. So the check would go along with `!nodes.Has(*endpoint.NodeName)`

**Review by jcaamano — COMMENTED** [review]
> I was a bit afraid of taking this route. Can it wait? 

Yup

**Review by jcaamano — COMMENTED** [review]
> Good idea! I made `getNodeSwitchTargetIPs` handle v4 and v6 at the same time. Does it look ok?
Also, do you prefer a free method because we're not really modifying `c *lbConfig`?  

I preferred the free method because it didn't make sense to pass as arguments data that is already available through lbConfig. I thought about keeping the receiver and removing the extra arguments but something that I don't remember know didn't make sense as well. So I suggested the free method.

**Review by jcaamano — COMMENTED** [review]
> @jcaamano as discussed, I implemented the new logic in one function, `getEndpointsForService`, that gets called once for a service and whose output is then parsed by the rest of the services controller to retrieve cluster-wide and per-node endpoint addresses. PTAL :)  

Looks good, added some comments to polish. CI looks bad though, hopefuly something else.

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/lb_config.go:888]: Yeah, I can also just check against FQDN. I was checking against known values (ipv4, ipv6) because I'm a bit paranoid about future implementation of `AddressType`, but let's go for the simpler solution :)

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/lb_config.go:877]: For `ports`? :-o

**Comment by ricky-rav** [comment]
> CI looks bad though, hopefuly something else.
I broke something... let me try to fix it :)

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/lb_config.go:955]: nit: this is a superfluous check, no problem on iterating on an empty map

**Review by jcaamano — COMMENTED** [review]
Everything looks good @ricky-rav.

The smal nit, up to you to act on it, and just that failing lane.

Also, woudl you be interested in running this dowsntream to check on the disruption results before merging here?

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/lb_config.go]: right, I'll move the check to the log line further below... no need to print the local endpoints if we don't need them in the first place.

**Comment by ricky-rav** [comment]
> Also, would you be interested in running this dowsntream to check on the disruption results before merging here?

Yes! Let's see how it goes here: https://github.com/openshift/ovn-kubernetes/pull/2109

File: go-controller/pkg/ovn/controller/services/lb_config.go
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/lb_config.go:975]: nit: double spacing in this log and the one above

**Comment by ricky-rav** [comment]
Our downstream CI fails on the `e2e-metal-ipi-ovn-ipv6` job on tests related to services and endpoints:

- [sig-network] Networking Granular Checks: Services should function for endpoint-Service: http [Suite:openshift/conformance/parallel] [Suite:k8s]
- [sig-network] Networking Granular Checks: Services should function for endpoint-Service: udp [Suite:openshift/conformance/parallel] [Suite:k8s]
- [sig-network] Services should be able to connect to terminating and unready endpoints if PublishNotReadyAddresses is true [Suite:openshift/conformance/parallel] [Suite:k8s]
- [sig-network] Networking Granular Checks: Services should function for multiple endpoint-Services with same selector [Suite:openshift/conformance/parallel] [Suite:k8s] 
- [sig-network] Services should be able to update service type to NodePort listening on same port number but different protocols [Suite:openshift/conformance/parallel] [Suite:k8s]
- [sig-network] Services should fallback to terminating endpoints when there are no ready endpoints with externallTrafficPolicy=Cluster [Suite:openshift/conformance/parallel] [Suite:k8s]

https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_ovn-kubernetes/2109/pull-ci-openshift-ovn-kubernetes-master-e2e-metal-ipi-ovn-ipv6/1777348680731332608

These jobs are either not run in our upstream CI for IPv6 jobs or never run at all :( 

I'm trying to reproduce these failures on KIND and figure out how the tests get skipped...

**Comment by ricky-rav** [comment]
With the second fix, `e2e-metal-ipi-ovn-ipv6` is finally green.
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_ovn-kubernetes/2109/pull-ci-openshift-ovn-kubernetes-master-e2e-metal-ipi-ovn-ipv6/1779160887962112000

I'll add the code here.

**Comment by ricky-rav** [comment]
Failure in CI is on `[FAIL] Services of type NodePort [It] should listen on each host addresses`, which is a known issue https://github.com/ovn-org/ovn-kubernetes/issues/4275

**Review by jcaamano — APPROVED** [review]

**Pull request merged**
